lib/BoundaryValueDiffEqAscher/src/utils.jl

```diff
@@ -0,0 +1,225 @@
+function build_almost_block_diagonals(
+        zeta::Vector{T}, ncomp::I, mesh, ::Type{T}) where {T, I}
+    lside = 0
+    ncol = 2 * ncomp
+    n = length(mesh) - 1
+    # build integs (describing block structure of matrix)
+    rows = Vector{I}(undef, n)
+    cols = repeat([ncol], n)
+    lasts = repeat([ncomp], n)
+    let lside = 0
+        for i in 1:(n - 1)
+            lside = first(findall(x::Float64 -> x > mesh[i], zeta)) - 1
+            rows[i] = ncomp + lside
+        end
+    end
+    lasts[end] = ncol
+    rows[end] = ncol
+    g = IntermediateAlmostBlockDiagonal([zeros(rows[i], cols[i]) for i in 1:n], lasts)
+    return g
+end
+
+function __get_bcresid_prototype(prob::BVProblem, u)
+    return __get_bcresid_prototype(prob.problem_type, prob, u)
+end
+function __get_bcresid_prototype(::TwoPointBVProblem, prob::BVProblem, u)
+    prototype = if prob.f.bcresid_prototype !== nothing
+        prob.f.bcresid_prototype.x
+    else
+        first(prob.f.bc)(u, prob.p), last(prob.f.bc)(u, prob.p)
+    end
+    return prototype, size.(prototype)
+end
+function __get_bcresid_prototype(::StandardBVProblem, prob::BVProblem, u)
+    prototype = prob.f.bcresid_prototype !== nothing ? prob.f.bcresid_prototype : zero(u)
+    return prototype, size(prototype)
+end
+
+# Custom pivot LU factorization and substitution for simple usage and
+# customized array types
```

Comment:
This doesn't seem so good, it's not blocking based on cache sizes? What's the reason for this?

Reply:
The blocking factorization is used in the whole global collocation matrix(describe the discretized BVDAE system), but the simple pLU here is actually used for parameter condensation of each interior collocation matrix block.